### PR TITLE
Document module naming conventions

### PR DESCRIPTION
**Pull request by perazz**
This PR writes documentation for module naming conventions as implemented in FPM at
https://github.com/fortran-lang/fpm/pull/828

I'm opening it on the main branch as creating a new branch would break the automated pages build.

Check out the updated site here:
https://perazz.github.io/fpm-docs/en/spec/naming.html

cc @fortran-lang/admins @arteevraina @henilp105 @minhqdao 

I think the rules are good enough for now, so @everythingfunctional @certik I'd suggest to merge both PRs and move forward if you agree!

### CODE REVIEW

**Review by everythingfunctional — APPROVED** [review]
:+1:

File: pages/spec/naming.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
+# Module name requirements
+
+:::{note}
+Module naming requirements only apply to packages which are uploaded to a fpm registry; by default, no naming rules are enforced for local fpm projects.
```
Comment (certik) [inline pages/spec/naming.md:4]: Why not enforcing it for local packages too? We want this to become a habit.

File: pages/spec/naming.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+In order to reduce the chance of name collisions, ``all Fortran modules`` in all packages of a registry are required to begin with their package name.
+The following rules apply:
+
+### Module names
+
+1. Must be valid Fortran names (see above).
+2. Must begin with their package name
+3. Having the exact same name as the package is allowed;
+4. If longer, the ``default separator`` `__` between the package name chunk and what follows must be used.
```
Comment (certik) [inline pages/spec/naming.md:51]: What does "longer" mean?

File: pages/spec/naming.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+These are non-mandatory styling suggestions to improve code readability and uniformity.
+:::
+
+It's recommended that the public API of each package is contained in a top-level module, whose name is same as the package name.
+For example, assuming a package ``DateTime`` deals with time and date in Fortran, one could have several modules deal with parts of it:
+
+```{code-block} fortran
+   module datetime__dates ; end module
+   module datetime__time  ; end module
+   module datetime__julian; end module
```
Comment (certik) [inline pages/spec/naming.md:146]: This should be just single underscore?

**Comment by certik** [comment]
Thanks for writing this up. I have quite a few comments:

* big one: I think we should enforce this for all packages; we of course make it easy to override/disable, but by default I think this should be enforced.
* let's first focus on what we recommend, only then list the corner cases (for example I think most package names will not have underscores in their names, so we should move `my_pkg` example into some corner case and focus one package names such as `datetime`)
* Should we do two underscores or one? It seems using two underscores is kind of long.

@milancurcic mentioned a use case where he doesn't even want to use an underscore I think.

Let's use this doc to iterate on the design that we can agree on.

File: pages/spec/naming.md (hunk 1 above)
Comment (milancurcic) [reply pages/spec/naming.md:4]: Agree with Ondrej here.

**Review by milancurcic — COMMENTED** [review]
Thanks for working on this. Beside the specific comments, the double underscore requirement seems the only thing that stands out as awkward. I know that there was a discussion about it but I haven't read it completely nor carefully and will do that. I can't imagine most people wanting to put double underscores in their module names.

File: pages/spec/naming.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+   mypackage123   ! Numbers OK
+```
+
+*Examples of invalid package names:*
+
+```fortran
+  my__package       ! Contains package__module separator
+  package__         ! Contains separator
+  package_          ! Ends with underscore
+  my pac$age        ! Spaces and all symbols besides `_` not allowed
```
[the remark this answers is not in the document]
Comment (certik) [reply pages/spec/naming.md:106]: I didn't know we can use `-`. We should document it right here! :)

File: pages/spec/naming.md (hunk 1 above)
Comment (perazz) [reply pages/spec/naming.md:4]: The reason is that enforcing them now they'd break everything, starting from building fpm (toml-f does not comply). Maybe we should just warn folks for a while, to give them time to update their packages?

**Comment by perazz** [comment]
Thank you both @certik @milancurcic, this is very useful. The double underscore came up in https://github.com/fortran-lang/fpm/pull/828 as @everythingfunctional rightfully observed that a single underscore may lead to collisions relatively easily. For example: 

```
module date_time !  from package "date", handles time conversions
module date_time ! main module from package "date_time"
```

Things will get creepier as we're also allowing, like @milancurcic remembered, package names to have dashes (`-`), which will be converted to underscores (`_`) in all modules & fpm handling. 

So if I have a package named `fpm-regex` I can have modules named `fpm_regex` which could have several potential collisions with other modules e.g. from the `fpm` package.

That's the reasoning behind the double underscore: that ideally it would limit potential nontrivial collisions. But I 100% agree with you guys that using a single one would be far more elegant. So, what should we do?

File: pages/spec/naming.md (hunk 2 above)
Comment (perazz) [reply pages/spec/naming.md]: That is not just the package name. Agreed: ugly sentence, I removed it

File: pages/spec/naming.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+## FPM registry names: rules for packages and modules
+
+In order to reduce the chance of name collisions, ``all Fortran modules`` in all packages of a registry are required to begin with their package name.
+The following rules apply:
+
+### Module names
+
+1. Must be valid Fortran names (see above).
```
[the remark this answers is not in the document]
Comment (perazz) [reply pages/spec/naming.md]: Makes sense, removed this item.

File: pages/spec/naming.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+2. Must begin with their package name
+3. Having the exact same name as the package is allowed;
```
[the remark this answers is not in the document]
Comment (perazz) [reply pages/spec/naming.md]: Removed

File: pages/spec/naming.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (perazz) [reply pages/spec/naming.md:106]: Done in https://github.com/fortran-lang/fpm-docs/pull/90/commits/1bc2cfd4bd02a839e232b53b58620738249a9834

**Comment by perazz** [comment]
Major takeaways from last week's [Fortran Monthly Call](https://fortran-lang.discourse.group/t/fortran-monthly-call-january-2023/5001/9):

- We should begin enforcing namespacing asap, to let people adapt sooner rather than later. Do not do that too gradually: people don't want to keep changing their builds
- Double underscore `__` is awkward, are there alternatives which are less strict but more ergonomic i.e. use single underscores `_` only? 
- Name enforcing should be an option in `fpm.toml`
- We should not confuse package namespacing with module namespacing

Based on the *current* implementation, to build fpm one would have to: 
- `M_CLI2` is OK: module and package have the same name
- `fpm` is not OK: modules need to be renamed, e.g. `fpm_manifest` -> `fpm__manifest`
- `toml-f` is not OK: need double underscore + the prefix does not match the package name. e.g. `tomlf_constants` -> `toml_f__constants`

So yeah lots of changes need be made. How about specifying a user-defined prefix for each package in `fpm.toml`? It would unlink the package name from the module names, so, not necessarily a good idea. But it would enable back using the single underscore, and the registry could easily check there are no duplicate prefixes. For example, assume on package `toml-f`: 

```
[build]
module-naming = "tomlf"
```

specifies a custom prefix; would enable more readable names to be used
```
module tomlf_ser
module tomlf_constants
[...]
```  

while 
```
[build]
module-naming = true
```

just goes back to the default conventions: 
```
module toml_f__ser
module toml_f_constants
```

I don't think doing this is too much work on the fpm side, but I haven't thought well enough if enabling yet another keyword to float around in the registry would cause more harm than good.

**Comment by everythingfunctional** [comment]
I think part of the answer to allowing "custom prefixes" should be based on how difficult will it be for the registry to still help prevent upload of packages with conflicting names. Would it be allowed to upload a new version of a package with a different custom prefix? Can a package use both? I.e.

```fortran
module veggies ! entry point
...
module vg_test ! abbreviated prefix
```

**Comment by perazz** [comment]
Let's assume the registry can do all those checks for now. A custom prefix has basically the same role as a package name, but if we prescribe slightly different rules: 1) no underscore or dashes, 2) followed by single underscore, it belongs to a second namespace and does not steal further package names

In your example, we could have package `veggies` with prefix `vg`:
```
module veggies        ! entry point
module vg_utils       ! custom prefix
module veggies__utils ! standard prefix
```

get along just fine with package `vg` (assume custom prefix `v`): 
```
module vg          ! entry point
module vg__utils   ! standard namespaced version
module v_utils     ! with custom prefix "v"
```

**Comment by everythingfunctional** [comment]
That seems to work well enough.

**Comment by milancurcic** [comment]
> Based on the current implementation, to build fpm one would have to:
> 
> * `M_CLI2` is OK: module and package have the same name
> * `fpm` is not OK: modules need to be renamed, e.g. `fpm_manifest` -> `fpm__manifest`
> * `toml-f` is not OK: need double underscore + the prefix does not match the package name. e.g. `tomlf_constants` -> `toml_f__constants`

It seems like it will boil down to:

* For single-module packages, those packages whose modules have a different name from that of the package will need to be changed; others are OK.
* For multi-module packages, most if not all will need to be changed due to the double-underscore requirement. That alone makes it a non-starter IMO.

So perhaps our survey to do is to assume the single-underscore requirement and see how many packages break with that (in addition to the package name prefix rule).

As an aside, a project of mine may be an interesting case here. [datetime-fortran](https://github.com/wavebitscientific/datetime-fortran) is a single-module package with the name `datetime` (rationale: original repo name datetime-fortran has "fortran" in it for branding purpose; however, it seemed redundant to include "fortran" when it came to naming it as an fpm package). The single module is named `datetime_module` (instead of just `datetime`) to avoid a name clash with the type `datetime` that it exports.

If we require a single-module package to have the package and module names the same, `datetime` fails the check, and I would need to either rename the package or the derived type.

Have we discussed yet whether the module name must be the same as the (single-module) package name, or is it rather a recommendation?

**Comment by perazz** [comment]
> Have we discussed yet whether the module name must be the same as the (single-module) package name, or is it rather a recommendation?

Differentiating between 1-module and n-module packages is an unnecessary restriction imho. Like you’re saying, if you have colliding type names you can always  add a suffix to the main module like `datetime_module`.

let me analyze your case: package is named `datetime-fortran` in fpm.toml and `datetime` is reserved by a derived type. 

You can:
1) use standard naming and call it `module datetime_fortran`
2) add `module-naming=“datetime”` you fpm toml and you can keep using `module datetime_module`. That’s it! The caveat is that your reserved keyword has no dashes or underscores. 

I’m starting to like this idea of the two options. Now though, fpm should be able to query all custom prefixes from the registry, it’s more complicated than just checking that there are no duplicate package names. And we should think of cases like “user wants to change the prefix”

**Comment by perazz** [comment]
I believe the implementation is completed, see [updated docs](https://perazz.github.io/fpm-docs/en/spec/naming.html) page. 
Like discussed, the user now has 3 options: 
- No rules enforced: backward-compatible with no changes, or set `module-naming=false` in fpm.toml
- Standardized naming: package+double underscore: `package_name__module_name`, choose using `module-naming=true` in fpm.toml
- Custom prefix: prefix(no symbols) + underscore: `mypkg_module_name`, choose using `module-naming="mypkg"` in fpm.toml

I've implemented these rules in [fpm](https://github.com/fortran-lang/fpm/pull/828), which now also reads in the rules and checks all options in all dependencies. I think it's basically ready to merge.

**Comment by awvwgk** [comment]
Merged since we now support this in fpm.

**Pull request merged**